File: iPhone/VerBank-CSTSv3-IOSClientApi/ClientStation/api/event/temp/API_IDEvent_EnquirerContainer.m
////
////  API_IDEvent_EnquirerContainer.m
////  JEDIv7-CSTSv3-ClientAPI
////
////  Created by baolin on 13/5/15.
////  Copyright (c) 2013年 wangyubo. All rights reserved.
////
//
//#import "API_IDEvent_EnquirerContainer.h" 
//
//id containerLock=@"lock";
// 
//@implementation API_IDEvent_EnquirerContainer
//
//-(id)init{
//    if(self=[super init])
//    {
//        listenerListMapByEventName=[[NSMutableDictionary alloc]init];
//        eventEnquireMapByListener=[[NSMutableDictionary alloc]init];
//    }
//    return self;
//}
//
//-(NSMutableArray *)getEnquiredListeners:(NSString *)eventName{
//    @synchronized(containerLock){
//        if ([listenerListMapByEventName isEqual:eventName]) {
//            NSMutableArray *array=[[NSMutableArray alloc]init];
//            NSEnumerator * enumeratorValue = [listenerListMapByEventName objectEnumerator];
//            for (API_IDEventListener *object in enumeratorValue) {
//                [array addObject:object];
//            }
//            return array;
//        }else{
//           return [[NSMutableArray alloc]init];
//        }
//    
//    }
//}
//
//-(void)addListener:(API_IDEventListener *)listener eventName:(NSString *)eventName{
//    @synchronized(containerLock){
//        API_IDEvent_Enquirer *see=nil;
//        if ([eventEnquireMapByListener isEqual:eventName]) {
//            see=[eventEnquireMapByListener objectForKey:listener];
//        }else{
//            see=[[API_IDEvent_Enquirer alloc]init:listener];
//            [eventEnquireMapByListener setObject:see forKey:[listener copy]];
//        }
//        
//    if (![see containsEvent:eventName]) {
//        [see addEventName:eventName];
//        NSMutableArray *listenerlist=nil;
//        if ([listenerListMapByEventName isEqual:eventName]) {
//            listenerlist=[listenerListMapByEventName objectForKey:eventName];
//        }else{
//            listenerlist=[[NSMutableArray alloc]init];
//            [listenerListMapByEventName setObject:listenerlist forKey:eventName];
//        }
//        }
//    }
//}
//
//-(void)removeListener:(API_IDEventListener *)listener eventName:(NSString *)eventName{
//    @synchronized(containerLock){
//        if (![eventEnquireMapByListener isEqual:listener]) {
//            return;
//        }
//    API_IDEvent_Enquirer *see=[eventEnquireMapByListener objectForKey:listener];
//    if(![see containsEvent:eventName]){
//        return;
//    }
//    [see removeEventName:eventName];
//   [[listenerListMapByEventName objectForKey:eventName]removeObject:listener];
//}
//}
//
//-(void)removeListener:(API_IDEventListener *)listener{
//    @synchronized(containerLock){
//        if (![eventEnquireMapByListener isEqual:listener]) {
//            return;
//        }
//        API_IDEvent_Enquirer *see=[eventEnquireMapByListener objectForKey:listener];
//        [eventEnquireMapByListener removeObjectForKey:listener];
//        NSArray *eventList=[see getEventNameList];
//        for (int i=0; i<[eventList count]; i++) {
//            NSString *eventName=[eventList objectAtIndex:i];
//            [[listenerListMapByEventName objectForKey:eventName]removeObject:listener];
//        }
//    }
//}
//@end

File: WeChatService/EX02-CSTSv3-ClientAPI/src/jedi/ex01/client/station/api/event/API_IDEvent_EnquirerContainer.java
package jedi.ex01.client.station.api.event;

import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class API_IDEvent_EnquirerContainer {
	private HashMap<String, LinkedList<API_IDEventListener>> listenerListMapByEventName = new HashMap<String, LinkedList<API_IDEventListener>>();
	private HashMap<API_IDEventListener, API_IDEvent_Enquirer> eventEnquirerMapByListener = new HashMap<API_IDEventListener, API_IDEvent_Enquirer>();
	private Object containerLock = new Object();

	public List<API_IDEventListener> getEnquiredListeners(String eventName) {
		synchronized (containerLock) {
			if (listenerListMapByEventName.containsKey(eventName)) {
				return Collections.unmodifiableList(listenerListMapByEventName
						.get(eventName));
			} else {
				return new LinkedList<API_IDEventListener>();
			}
		}
	}

	public void addListener(API_IDEventListener listener, String eventName) {
		synchronized (containerLock) {
			API_IDEvent_Enquirer see = null;
			if (eventEnquirerMapByListener.containsKey(listener)) {
				see = eventEnquirerMapByListener.get(listener);
			} else {
				see = new API_IDEvent_Enquirer(listener);
				eventEnquirerMapByListener.put(listener, see);
			}
			if (!see.containsEvent(eventName)) {
				see.addEventName(eventName);
				LinkedList<API_IDEventListener> listenerList = null;
				if (listenerListMapByEventName.containsKey(eventName)) {
					listenerList = listenerListMapByEventName.get(eventName);
				} else {
					listenerList = new LinkedList<API_IDEventListener>();
					listenerListMapByEventName.put(eventName, listenerList);
				}
				listenerList.add(listener);
			}
		}
	}

	public void removeListener(API_IDEventListener listener, String eventName) {
		synchronized (containerLock) {
			if (!eventEnquirerMapByListener.containsKey(listener)) {
				return;
			}
			API_IDEvent_Enquirer see = eventEnquirerMapByListener.get(listener);
			if (!see.containsEvent(eventName)) {
				return;
			}
			see.removeEventName(eventName);
			listenerListMapByEventName.get(eventName).remove(listener);
			if (see.eventSize() == 0) {
				removeListener(listener);
			}
		}
	}

	public void removeListener(API_IDEventListener listener) {
		synchronized (containerLock) {
			if (!eventEnquirerMapByListener.containsKey(listener)) {
				return;
			}
			API_IDEvent_Enquirer see = eventEnquirerMapByListener
					.remove(listener);
			List<String> eventList = see.getEventNameList();
			Iterator<String> it = eventList.iterator();
			while (it.hasNext()) {
				String eventName = it.next();
				listenerListMapByEventName.get(eventName).remove(listener);
			}
		}
	}

}
